packages/backend-app-api/src/services/implementations/discovery/discoveryServiceFactory.ts

```diff
@@ -18,15 +18,65 @@ import {
   coreServices,
   createServiceFactory,
 } from '@backstage/backend-plugin-api';
-import { HostDiscovery } from './HostDiscovery';
+import { MultipleBackendHostDiscovery } from './MultipleBackendHostDiscovery';
+import { InstanceRegistration } from './InstanceRegistration';
+import { createRouter } from './service/router';
 
 /** @public */
 export const discoveryServiceFactory = createServiceFactory({
   service: coreServices.discovery,
   deps: {
     config: coreServices.rootConfig,
+    rootFeatureRegistry: coreServices.rootFeatureRegistry,
+    lifecycle: coreServices.rootLifecycle,
+    tokenManager: coreServices.tokenManager,
+    rootHttpRouter: coreServices.rootHttpRouter,
+    rootLogger: coreServices.rootLogger,
   },
-  async factory({ config }) {
-    return HostDiscovery.fromConfig(config);
+  async createRootContext({
+    config,
+    lifecycle,
+    rootFeatureRegistry,
+    rootHttpRouter,
+    rootLogger,
+    tokenManager,
+  }) {
+    const logger = rootLogger.child({
+      service: 'discovery',
+    });
+    const discovery = MultipleBackendHostDiscovery.fromConfig(config, {
+      rootFeatureRegistry,
+      tokenManager,
+      logger,
+    });
+    lifecycle.addStartupHook(() => {
+      if (!discovery.isGateway) {
+        const registration = InstanceRegistration.fromConfig(config, {
+          rootFeatureRegistry,
+          discovery,
+          tokenManager,
+        });
+
+        registration.startHeartbeat();
+      } else if (discovery.isGateway && !discovery.isInitialized) {
+        discovery.initialize();
+      }
+    });
+
+    if (discovery.isGateway) {
+      rootHttpRouter.use(
+        '/api/discovery',
+        createRouter({
```

Comment:
Tbh I think we'll want this to be an actual plugin that provides an API instead, rather than something registered as a root service.

Reply:
Makes sense, I was flipping back and forth between plugin and root service. It's a bit of a circular dependency if we have the discovery API and a discovery plugin, but I think there's an easy way through that.